Posts/2021/11/tfl-formatenumeration.md

```diff
@@ -0,0 +1,150 @@
+PS > Get-Process -Name PWSH | Format-Table -Property ProcessName, Modules
+
+ProcessName Modules
+----------- -------
+pwsh        {System.Diagnostics.ProcessModule (pwsh.exe), 
+             System.Diagnostics.ProcessModule (ntdll.dll),
+             System.Diagnostics.ProcessModule (KERNEL32.DLL),
+             System.Diagnostics.ProcessModule (KERNELBASE.dll),
+             System.Diagnostics.ProcessModule (apphelp.dll),
+             System.Diagnostics.ProcessModule (USER32.dll),
+             System.Diagnostics.ProcessModule (win32u.dll),
+             System.Diagnostics.ProcessModule (GDI32.dll)…}   
+```
+
+In the above output, you can see output for eight modules.
+In writing this, there are actually 239 actual modules for the PowerShell process.
+If you need to see all the modules, you could set `$FormatEnumerationLimit` to a larger number (e.g. 999) in the shell.
+Alternatively, if you set `$FormatEnumerationLimit` to -1, PowerShell displays all occurrences, which may be more than you want in most cases!
+I set the limit to 99 in my profile file and that is usually more than sufficient.
+
+## Scoping of $FormatEnumerationLimit
+
+One interesting thing I found is that `$FormatEnumerationLimit` is scoped differently to my expectations.
+If you use a format command within a function/script (i.e. a child of the global scope), the command only uses the value from the global scope.
+
+The following code contains a function to illustrate the issue:
+
+```powershell
+function Test-FormatLimit1
+{
+  # Change format enum limit
+  "In Function, limit is: [$FormatEnumerationLimit]"
+  $FormatEnumerationLimit = 1
+  "After changing: [$FormatEnumerationLimit]"
+  Get-Process | Select-Object -Property Name, Threads -First 4
+}
+```
+
+You might think that this code would display the first thread in each of the first four processes. 
+You might, but you would be wrong, as you can see here:
+
+```powershell-console
+PS> # Here show the value and call the functin
+PS> "Before calling: [$FormatEnumerationLimit]"
+Before calling: [4]
+PS> Test-FormatLimit1
+In Function, limit is: [4]
+After changing: [1]
+
+Name                    Threads
+----                    -------
+AggregatorHost          {5240}
+ApplicationFrameHost    {16968, 2848}
+AppVShNotify            {9164}
+Atom.SDK.WindowsService {4064, 4908, 4912, 19144…}
+```
+
+As you can see from this output, the final process shows FOUR threads not ONE.
+This is because PowerShell seems to only use the globally scoped value, not the locally scoped copy.
+To get around this curious scoping, you can re-write the function like this:
+
+```powershell
+
+function Test-FormatLimit2
+{
+  # Change format enum limit Globally
+  $Old = $Global:FormatEnumerationLimit
+  $Global:FormatEnumerationLimit = 2
+  "After changing: [$Global:FormatEnumerationLimit]"
+  Get-Process | Select-Object -Property Name, Threads -First 4
+  # Change it back
+  $Global:FormatEnumerationLimit = $Old
+}
```

Comment:
Make the examples symmetrical. Use the same limits in both examples. Only change the scope. Don't want to confuse folks by changing multiple parameters.

```suggestion
function Test-FormatLimitGlobal
{
  # Change format enum limit Globally
  $Old = $Global:FormatEnumerationLimit
  $Global:FormatEnumerationLimit = 1
  "After changing: [$Global:FormatEnumerationLimit]"
  Get-Process | Select-Object -Property Name, Threads -First 4
  # Change it back
  $Global:FormatEnumerationLimit = $Old
}
```